src/libraries/CalldataDecoder.sol

```diff
@@ -3,11 +3,101 @@ pragma solidity ^0.8.23;
+
+    /// @notice mask used for offsets and lengths to ensure no overflow
+    /// @dev no sane abi encoding will pass in an offset or length greater than type(uint32).max
+    ///      (note that this does deviate from standard solidity behavior and offsets/lengths will
+    ///      be interpreted as mod type(uint32).max which will only impact malicious/buggy callers)
+    uint256 constant OFFSET_OR_LENGTH_MASK = 0xffffffff;
+    uint256 constant OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN = 0xffffffe0;
+
+    /// error SliceOutOfBounds();
+    uint256 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+
     /// @notice Removes the selector from the calldata and returns the encoded params.
     function removeSelector(bytes calldata data) internal pure returns (bytes calldata params) {
-        assembly {
+        assembly ("memory-safe") {
+            if lt(data.length, 4) {
+                mstore(0, SLICE_ERROR_SELECTOR)
+                revert(0x1c, 4)
+            }
             params.offset := add(data.offset, 4)
             params.length := sub(data.length, 4)
         }
     }
+
+    /// @notice Decode the signature and hook data from the calldata
+    /// @dev The calldata is expected to be encoded as `abi.encode(bytes signature, bytes hookData)`
+    function decodeSignatureWithHookData(bytes calldata data)
+        internal
+        pure
+        returns (bytes calldata signature, bytes calldata hookData)
+    {
+        signature = toBytes(data, 0);
+        hookData = toBytes(data, 1);
+    }
+
+    /// @notice Decode the keyHash, signature, and hook data from the calldata
+    /// @dev The calldata is expected to be encoded as `abi.encode(bytes32 keyHash, bytes signature, bytes hookData)`
+    function decodeSignatureWithKeyHashAndHookData(bytes calldata data)
+        internal
+        pure
+        returns (bytes32 keyHash, bytes calldata signature, bytes calldata hookData)
+    {
+        assembly {
+            keyHash := calldataload(data.offset)
+        }
+        signature = toBytes(data, 1);
+        hookData = toBytes(data, 2);
+    }
+
+    /// @notice Decode the signature, appSeparator, contentsHash, and contentsDescr from the calldata
+    ///         the return values MUST be checked for length before use
+    /// @dev The calldata is expected to be encoded as `abi.encode(bytes signature, bytes32 appSeparator, bytes32 contentsHash, string contentsDescr)`
+    ///      there may be an uint16 contentsLength at the end of the calldata, but this is not used
+    function decodeTypedDataSig(bytes calldata data)
+        internal
+        pure
+        returns (bytes calldata signature, bytes32 appSeparator, bytes32 contentsHash, string calldata contentsDescr)
+    {
+        signature = toBytes(data, 0);
+        assembly {
+            appSeparator := calldataload(add(data.offset, 0x20))
+            contentsHash := calldataload(add(data.offset, 0x40))
+        }
+        bytes calldata contentsDescrBytes = toBytes(data, 3);
```

Comment:
I think you can just cast to a string and not declare the bytes variable.. but def test to make sure

Reply:
Ah yep I think it should work?